cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/SparkPipelineRunner.java

```diff
@@ -378,6 +356,123 @@ public void runPipeline(PipelinePhase pipelinePhase, String sourcePluginType,
     }
   }
 
+  private SparkCollection<Object> handleAutoJoin(AutoJoiner autoJoiner, AutoJoinerContext autoJoinerContext,
+                                                 Map<String, SparkCollection<Object>> inputDataCollections) {
+    JoinDefinition joinDefinition = autoJoiner.define(autoJoinerContext);
+
+    // join required stages first to cut down the data as much as possible
+    List<JoinStage> requiredStages = joinDefinition.getStages().stream()
+      .filter(s -> s.isRequired())
+      .collect(Collectors.toList());
+    List<JoinStage> optionalStages = joinDefinition.getStages().stream()
+      .filter(s -> !s.isRequired())
+      .collect(Collectors.toList());
+    List<JoinStage> orderedStages = new ArrayList<>(requiredStages.size() + optionalStages.size());
```

Comment:
Can you explain a bit why we want an ordered stages? From the logic below, we are basically going through the list one by one, and then determine the join type based on left is required or right is required. Why do we want to loop through required stages first?

Reply:
there's a comment above but I can expand on it. If there is an outer join and an inner join, it is more performant to do the inner join first because it will generate less intermediate data.

Reply:
though now that I think more about it, it may be better not to do this, and just follow the order provided by the plugin.

Reply:
actually, going to re-insert the sorting. It is a good amount simpler to perform inner joins followed by left outer joins rather than inter mixing everything.